### PR TITLE
ANother attempt to use moveToCodePointOffset for cursor routing

### LINKED ISSUE #10960: Braille: Variation Selectors break cursor positions
### Steps to reproduce:
1. Consider the following line:
  > ⚠️ test
2. Move the cursor on the word (test) using routing cursors.

### Actual behavior:
The cursor is moved on — the desired position + 1 —. Also ⚠️ takes 2 braille cells.

### Expected behavior:
The cursor should be moved on the desired position. Also ⚠️ symbol should takes one cell only.
With wordpad and in form mode in Firefox, cursor is properly moved.

#### NVDA installed/portable/running from source:
Installed and portable

#### NVDA version:
2019.3, 2020.1

#### Windows version:
10 Insider (64-bit) build 19592.1001

#### Name and version of other software in use when reproducing the issue:
notepad, Microsoft Word, Firefox (in browse mode), Run dialog (windows+r), etc..

#### Does the issue still occur after restarting your computer?
Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
No

#### If addons are disabled, is your problem still occuring?
Yes

#### Did you try to run the COM registry fixing tool in NVDA menu / tools?
No

CC @leonardder

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Fixes #10960
Replaces #16477, #16497 

### Summary of the issue:
There are cases where moving one character on a textInfo instance actually moves more than one unicode point offset. This is described by @mltony in the doc string for `textInfos.TextInfo.moveToCodepointOffset`.
This causes of by one errors when cursor routing, since we're asking the textInfo to move by 1 characters, that might be presented by two or even more characters within the liblouis mapping.

### Description of user facing changes
Cursor routing should be more reliable.

### Description of development approach
@mltony's creation of `moveToCodepointOffset` allows us to move x code points from the start of the reading unit. As we're using 32 bit encoding for liblouis, every character as presented by liblouis is equal to one code point. Therefore we can safely assume that this method to move is much more reliable than the previous method.

#### Differences with #16477, #16497
To fix the issue mentioned in https://github.com/nvaccess/nvda/issues/10960#issuecomment-2115810057, i.e. with bullets in Word, I did the following:

1. Changed `moveToCodepointOffset` in the base textInfo to use a new method `_getTextForCodepointMovement` to get the text to operate on. this was the `text property before`.
2. `_getTextForCodepointMovement` defaults to `return self.text`, however for Word UIA, it uses getTextWithFields to construct the text to operate on. This ensures `moveToCodepointOffset` works with text in which bullets and numbering are excluded.

### Testing strategy:
- Test str from #10960
- Unit test that basically mimics the str from  #10960

### Known issues with pull request:
In textInfo instances where multiple codepoint emoji are treated as one character (e.g. in Word), `moveToCodepointOffset` raises an error when trying to move to the second code point. Therefore for routing, this should be handled more gracefully, that's why when `moveToCodePointOffset` fails, we move one character back with a maximum of 10 iterations. This behavior is covered by a unit test.

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

## Summary by CodeRabbit

- **New Features**
  - Enhanced text extraction in Word documents by stripping list bullets during character movement.
  - Improved braille cursor routing for Unicode variation selectors and decomposed characters.

- **Bug Fixes**
  - Enhanced error handling and logging for braille text information retrieval.

- **Tests**
  - Added new unit tests for braille cursor routing, including tests for emoji and composite characters.

- **Documentation**
  - Updated user documentation to highlight improvements in braille cursor routing reliability.

### CODE REVIEW

File: tests/unit/test_braille/test_routing.py [hunk 1]
```diff
@@ -147,3 +147,58 @@ def test_moveCaret_always_instantActivate(self):
...
+		region.routeTo(pos)
+		ti = obj.makeTextInfo(textInfos.POSITION_CARET)
+		ti.expand(textInfos.UNIT_CHARACTER)
+		self.assertEqual(ti.text, testText[1:4])
+		index = 3  # Position of ּ
+		pos = region.rawToBraillePos[index]
+		region.routeTo(pos)
+		ti = obj.makeTextInfo(textInfos.POSITION_CARET)
+		ti.expand(textInfos.UNIT_CHARACTER)
+		self.assertEqual(ti.text, testText[1:4])
```
[the remark this answers is not in the document]
Comment (LeonarddeR) [reply tests/unit/test_braille/test_routing.py:204]: @coderabbitai Do you have concrete suggestions on how to cover edge cases?

File: source/braille.py [hunk 2]
```diff
@@ -1426,14 +1426,35 @@ def update(self):
...
+				else:
+					logFunc = log.error
+					curPos = pos
+				logFunc(f"Error in moveToCodepointOffset in iteration {i + 1} (position {curPos}")
 		dest = self._readingInfo.copy()
 		dest.collapse()
-		# and move pos characters from there.
-		dest.move(textInfos.UNIT_CHARACTER, pos)
+		if curPos > 0:
+			dest.move(textInfos.UNIT_CHARACTER, curPos)
```
[the remark this answers is not in the document]
Comment (LeonarddeR) [reply source/braille.py:1457]: @coderabbitai logging the exception is not useful on debug level. I can change log.error to log.exception, that will always log on the error level with the exception attached.
Furthermore, we only need to catch RuntimeError, and the 10 iterations are intentional. That said, we could check the time at every iteration and ensure that this doesn't take longer than 0.5 second. Could be interesting if you could suggest an approach to 
Do you thinkhandle that. there's still something to suggest here?

File: source/textInfos/__init__.py [hunk 3]
```diff
@@ -803,7 +807,7 @@ def moveToCodepointOffset(
 			we reduce the count of characters in order to make sure
 			the algorithm makes some progress on each iteration.
 		"""
-		text = self.text
+		text = self._getTextForCodepointMovement()
```
[the remark this answers is not in the document]
Comment (LeonarddeR) [reply source/textInfos/__init__.py:810]: @coderabbitai this suggestion makes no sense. You're suggesting to call text on the copied textInfo instance, but that is definitely different from the text variable that's initialized before the loop.

File: tests/unit/test_braille/test_routing.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (LeonarddeR) [reply tests/unit/test_braille/test_routing.py:204]: This is lovely. Please make this a commitable suggestion.

[the remark this answers is not in the document]
Comment (LeonarddeR) [reply tests/unit/test_braille/test_routing.py:204]: I can't commit these. Please wrap them in a suggestion block so it will be actionable

[the remark this answers is not in the document]
Comment (LeonarddeR) [reply tests/unit/test_braille/test_routing.py:204]: I'm afraid this was still not what I intended, since this removed code
I fixed it in commit 7bc812653ca6fa20e363631efd6f779d4791a1d1 , please learn from it.

File: source/braille.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (LeonarddeR) [reply source/braille.py]: I changed stuff in 2fb01605186be3d37b6ea337ae31d7ca8677a785

[the remark this answers is not in the document]
Comment (LeonarddeR) [reply source/braille.py]: Is there anything you think I should still improve, taking my earlier feedback in https://github.com/nvaccess/nvda/pull/16876#discussion_r1682253100into account?

**Comment by michaelDCurran** [comment]
@LeonarddeR  can you mention in the PR description how this pr improves upon #16477 and #16497 , E.g. how it solves the MS Word bullet / numbering problem?

**Comment by LeonarddeR** [comment]
@michaelDCurran Sure, just did.

File: source/braille.py [hunk 4]
```diff
@@ -1426,14 +1426,39 @@ def update(self):
...
+		"""Fetches a collapsed TextInfo at the specified braille position in the region.
+		:param braillePos: The braille position.
+			If no textInfo could be found at braillePos,
+			try to find one at braillePos - 1 until a position has been found.
+		"""
 		pos = self._rawToContentPos[self.brailleToRawPos[braillePos]]
 		# pos is relative to the start of the reading unit.
-		# Therefore, get the start of the reading unit...
+		maxIterations = 10
+		start_time = time.time()
```
Comment (seanbudd) [inline source/braille.py:1437]:
```suggestion
		startTime = time.time()
```

File: tests/unit/test_braille/test_routing.py [hunk 5]
```diff
@@ -147,3 +147,58 @@ def test_moveCaret_always_instantActivate(self):
...
+		self.assertEqual(ti.text, testText[4])
+		region = braille.TextInfoRegion(obj)
+		region.update()
+		index = 1  # Position of ב
+		pos = region.rawToBraillePos[index]
+		region.routeTo(pos)
+		ti = obj.makeTextInfo(textInfos.POSITION_CARET)
+		ti.expand(textInfos.UNIT_CHARACTER)
+		self.assertEqual(ti.text, testText[1:4])
+		index = 3  # Position of ּ
```
Comment (seanbudd) [inline tests/unit/test_braille/test_routing.py:199]: position of?

**Review by seanbudd — COMMENTED** [review]
Looks good to me, just 2 minor issues

File: tests/unit/test_braille/test_routing.py (hunk 5 above)
Comment (LeonarddeR) [reply tests/unit/test_braille/test_routing.py:199]: > position of?

```suggestion
		index = 3  # Position of  ּ (\u5bc)
```

File: source/braille.py [hunk 6]
```diff
@@ -1426,14 +1426,39 @@ def update(self):
...
+				break
+			# Move curPos code points from the start.
+			# Note that, as liblouis uses 32 bit encoding internally,
+			# it is really safe to assume that one code point offset is equal to one character within liblouis.
+			# If an attempt fails, we try to move to the previous character
+			try:
+				return self._readingInfo.moveToCodepointOffset(curPos)
+			except RuntimeError:
+				msg = f"Error in moveToCodepointOffset in iteration {i + 1} (position {curPos}"
+				if i + 1 >= maxIterations or (exceeded := time.time() - start_time > 0.5):
```
Comment (LeonarddeR) [inline source/braille.py:1450]:
```suggestion
				if i + 1 >= maxIterations or (exceeded := time.time() - startTime > 0.5):
```

**Review by seanbudd — APPROVED** [review]

File: tests/unit/test_braille/test_routing.py (hunk 5 above)
Comment (seanbudd) [reply tests/unit/test_braille/test_routing.py]: Oh - I realized that this symbol is visually identical to my whitespace indicator for spaces, so this read to me as two spaces. Good reason to review with NVDA more! Thanks for adding the clarification, it's hard to see this character visually

Comment (LeonarddeR) [reply tests/unit/test_braille/test_routing.py]: It works both ways, good that you brought this up and that it is now clear to all parties, both sighted and sr users

**Comment by LeonarddeR** [comment]
I'm very puzzled about why this failed linting.
